File: Others/medium/MinNumberOfCoinsForChange.py
def minNumberOfCoinsForChange(n, denoms):

    num_coins = [float('inf') for i in range(n + 1)]

    num_coins[0] = 0

    for denom in denoms:
        for i in range(len(num_coins)):
            if i - denom >= 0:
                num_coins[i] = min(num_coins[i], num_coins[i - denom] + 1)

    return num_coins[n] if num_coins[n] != float('inf') else -1

File: Golang/medium/min-number-of-coins-for-change.go
package main

import "math"

func MinNumberOfCoinsForChange(n int, denoms []int) int {
	dp := make([]int, n+1)
	for idx := 0; idx <= n; idx++ {
		dp[idx] = math.MaxInt16
	}

	for _, denom := range denoms {
		if denom <= n {
			dp[denom] = 1
		}
	}
	dp[0] = 0

	for _, denom := range denoms {
		for idx := 1; idx <= n; idx++ {
			if idx-denom >= 0 && 1+dp[idx-denom] < dp[idx] {
				dp[idx] = 1 + dp[idx-denom]
			}
		}
	}

	if dp[n] == math.MaxInt16 {
		return -1
	}
	return dp[n]
}
